Documentation guide: Solve an equation algebraically

bertiewooster opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->
Solving sub-page: Solve an equation algebraically

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
First sub-page for solving main page, #23462 

#### Brief description of what is fixed or changed
Created solving folder under guides. Moved solving main page (formerly solving.md) into that folder, renamed to index.md. Put new page (Solve an equation algebraically) in that folder too.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

* other
  * A new, user-friendly page on solving an equation algebraically was added to the Guides section.
       
<!-- END RELEASE NOTES -->

bertiewooster committed:
Stub for first sub-page

bertiewooster committed:
Add examples for solve()
and equations that cannot be solved algebraically

bertiewooster committed:
Add issue-reporting sentence

bertiewooster committed:
Add solveset(), comparison to solve()

bertiewooster committed:
Add language

bertiewooster committed:
Merge branch 'sympy:master' into equation-algebraically

oscarbenjamin commented on doc/src/guides/solving/solve-equation-algebraically.md:
```diff
@@ -0,0 +1,90 @@
+# Solve an equation algebraically
+
+Use SymPy to solve an equation algebraically. For example, solving $x^2 = 4$ yields $x \in \{-2,2\}$.
```
This is not a good example because it would be better done in other ways e.g. by using a function specifically for finding polynomial roots.

Also this and many of the other examples give the impression that `solve` is for finding numeric solutions. It would be better to at least show an example that gives non-integer results like:
```python
In [4]: solve(x**2 + x - 1, x)
Out[4]: 
⎡  1   √5    √5   1⎤
⎢- ─ + ──, - ── - ─⎥
⎣  2   2     2    2⎦
```
Ideally you should emphasise the fact that we can solve equations involving symbols:
```python
In [5]: solve(x**2 - y, x)
Out[5]: [-√y, √y]
```
That's what SymPy's solve function can do that other functions like scipy's `fsolve` cannot do.

oscarbenjamin commented on doc/src/guides/solving/solve-equation-algebraically.md:
````diff
@@ -0,0 +1,90 @@
+[-2, 2]
+```
+
+3. Parse a string representing the equation into a form that SymPy can understand, then apply {func}`~.solve` to the parsed expression. 
+
+*But this doesn't work for equality; does for inequality. Is there some way to do this for equality?*
````
```
In [11]: solve(parse_expr("Eq(x**2, 1)"))
Out[11]: [-1, 1]
```

bertiewooster replied:
Thanks, that worked.

I also want to demonstrate parsing a string like "x**2 = y" -- that's the kind of expression that I originally used SymPy for, and I think a fairly likely use case. I tried using transformations=(convert_equals_signs) as @asmeurer suggested, but I'm having problems using it, or even importing transformations, as described in the [Parsing documentation](https://docs.sympy.org/dev/modules/parsing.html?highlight=parse_expr#sympy.parsing.sympy_parser.parse_expr). Any suggestions?

```
>>> solve(parse_expr("x**2 == y"), transformations=(convert_equals_signs))
Traceback (most recent call last):
  File "<pyshell#5>", line 1, in <module>
    solve(parse_expr("x**2 == y"), transformations=(convert_equals_signs))
NameError: name 'convert_equals_signs' is not defined
>>> transformations = (convert_equals_signs)
Traceback (most recent call last):
  File "<pyshell#6>", line 1, in <module>
    transformations = (convert_equals_signs)
NameError: name 'convert_equals_signs' is not defined
>>> from sympy.parsing.sympy_parser import transformations
Traceback (most recent call last):
  File "<pyshell#7>", line 1, in <module>
    from sympy.parsing.sympy_parser import transformations
ImportError: cannot import name 'transformations' from 'sympy.parsing.sympy_parser' (/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/sympy/parsing/sympy_parser.py)
```

oscarbenjamin replied:
Maybe something wrong with your installation. I can import it fine.

bertiewooster replied:
Hmm, weird. It works fine when I run it in Visual Studio Code. It didn't work in in IDLE, so maybe something was wrong with how SymPy got imported there.

bertiewooster replied:
Some progress: I can import transformations (I think it was a venv or conda issue).

In the following code, `parsed1` and `parsed2` work fine, but either of the last two lines, where I try to parse an expression containing an equals sign using `convert_equals_signs`, throws the exception `name 'x' is not defined`. I'm confused because I imported x from sympy.abc.

```
from sympy import parse_expr, solve
from sympy.abc import x
from sympy.parsing.sympy_parser import T, convert_equals_signs, transformations

parsed1 = parse_expr("2x", transformations=T[:6])
print(parsed1)
parsed2 = parse_expr("2 * x")
print(parsed2)
s = solve(parsed2, x)
print(s)

parsed3 = parse_expr("2 * x = 1", transformations=[convert_equals_signs])
parsed4 = parse_expr("2 * x = 1", transformations=T[9])
```

asmeurer replied:
You have to also include the standard_transformations (the default value of transformations). Something like `transformations=default_transformations + (convert_equals_signs,)` This whole thing is somewhat advanced, at least in its current state.

smichr replied:
Or you can use `transformations=T[1,9]` for both cases in parsed3-6.

oscarbenjamin commented on doc/src/guides/solving/solve-equation-algebraically.md:
````diff
@@ -0,0 +1,90 @@
+
+## Not all equations can be solved algebraically
+
+Some equations have no algebraic solution, in which case SymPy may return an empty set:
+
+```
+>>> from sympy import solve, Eq
+>>> from sympy.abc import x
+>>> eqn = Eq(3 * x - 7, 3 * x + 2)
+>>> solve(eqn)
+[]
````
The phrase "algebraic solution" is misleading. Here this equation does not have any solution full stop.

Other equations have solutions that cannot be expressed algebraically (symbolically) e.g.:
```python
In [12]: solve(cos(x)-x)
---------------------------------------------------------------------------
NotImplementedError
```
This equation has solutions (e.g. check `plot(cos(x), x)` to see the real one). We can find those numerically:
```python
In [14]: nsolve(cos(x)-x, x, 0)
Out[14]: 0.739085133215161
```
It is not however (as far as I know) possible to express the solution of this equation using any ordinary symbolic expression that anyone would recognise. There are different levels of what constitutes a solution expression:
https://en.wikipedia.org/wiki/Closed-form_expression#Analytic_expression

The `solve` function generally aims to compute solutions as explicit analytic expressions.

oscarbenjamin commented on doc/src/guides/solving/solve-equation-algebraically.md:
```diff
@@ -0,0 +1,90 @@
+
+{func}`~.solve`
+- is best for solving equations that...
+- produces outputs in the format of a Python list.
```
Actually it produces many different output formats:
```python
In [15]: solve(x**2 - 1)
Out[15]: [-1, 1]

In [16]: solve([x**2 - 1])
Out[16]: [{x: -1}, {x: 1}]

In [17]: solve([x**2 - 1], [x])
Out[17]: [(-1,), (1,)]

In [18]: solve([x**2 - 1], [x], set=True)
Out[18]: ([x], {(-1,), (1,)})

In [19]: solve({x**2 - 1}, [x])
Out[19]: [(-1,), (1,)]

In [20]: solve({x**2 - 1}, {x})
Out[20]: [{x: -1}, {x: 1}]

In [21]: solve([x**2 - 1, y**2 - 1], [x], set=True)
Out[21]: ([x], {(-1,), (1,)})

In [22]: solve([x**2 - 1, y**2 - 1], [x, y])
Out[22]: [(-1, -1), (-1, 1), (1, -1), (1, 1)]

In [23]: solve([x**2 - 1, y**2 - 1], [x, y], dict=True)
Out[23]: [{x: -1, y: -1}, {x: -1, y: 1}, {x: 1, y: -1}, {x: 1, y: 1}]
```
The key feature of `solve` is the fact that it produces explicit solution expressions that you can use to substitute into something else. That is its purpose and that is why the list of dicts output is the best so `dict=True` should always be used to get reliably useful output:
```python
In [24]: sols = solve([x**2 - 1, y**2 - 1], [x, y], dict=True)

In [25]: expr = sin(x)*exp(y)

In [26]: expr
Out[26]: 
 y       
ℯ ⋅sin(x)

In [27]: expr.subs(sols[0])
Out[27]: 
  -1       
-ℯ  ⋅sin(1)
```

bertiewooster replied:
Great, thanks. I incorporated your comment.

oscarbenjamin commented on doc/src/guides/solving/solve-equation-algebraically.md:
```diff
@@ -0,0 +1,90 @@
+
+so you may have to {func}`solve your equation numerically <sympy.solvers.solvers.nsolve>` instead.
+
+It is also possible that there is a way to solve your equation algebraically, and SymPy has not have implemented an appropriate algorithm. You can ask about this on the mailing list, or open an issue on GitHub. 
```
I think it's important to make clear here that the vast majority of arbitrary nonlinear equations are not analytically solvable. The classes of equations that are solvable are basically:

1. Linear equations.
2. Polynomials (except where limited by the Abel--Ruffini theorem).
3. Equations that can be solved by inverting some transcendental functions.
4. Problems that can be transformed into the cases above (e.g. by turning trig into polynomials).
5. A few other special cases that can be solved with something like the Lambert W function.

Pretty much everything else is not analytically solvable. This is not a limitation of SymPy but just a mathematical fact about which kinds of equations have solutions that can be expressed analytically in any way that could be considered useful.

This is something that it is important for users to understand. The reason it's important for them to understand this is because it very often means that they should not use the solve function because its whole purpose (analytic solutions) is something that is only achievable in very particular cases.

bertiewooster committed:
Notes from meeting with Aaron Meurer

bertiewooster committed:
Merge branch 'equation-algebraically' of github.com:bertiewooster/sympy into equation-algebraically

bertiewooster committed:
Suggested improvments, e.g.
solve equations involving symbols

bertiewooster committed:
Fixed example for parse_expr for an equation
(Still want to parse string like "x**2 = y")

bertiewooster committed:
Reformatted intro comparing solve() and solveset()
Gave example: why specify variable to solve for
Programmatically parsed a solve() solution
How to restrict to real solutions

bertiewooster commented:
Does anyone have an example where a mathematical solution exists to an equation, but SymPy returns an empty list because we haven't implemented an algorithm to find the solution?

oscarbenjamin commented:
> Does anyone have an example where a mathematical solution exists to an equation, but SymPy returns an empty list because we haven't implemented an algorithm to find the solution?

I don't immediately have one off hand but it should really be considered a bug. If there isn't an algorithm to find the solution then `NotImplementedError` should ideally be raised:
```python
In [3]: solve(x**2 - sin(x), x)
---------------------------------------------------------------------------
NotImplementedError: multiple generators [x, sin(x)]
No algorithms are implemented to solve equation x**2 - sin(x)
```
We should really improve that error message. I don't think many people will understand what generators are. In any case it should suggest alternatives like `nsolve`.

bertiewooster committed:
Tried to fix links in solving/index.md

bertiewooster committed:
Merge branch 'reorg-solving-docs' into equation-algebraically
Trying to get docs to build.

bertiewooster committed:
Show parsed, solved, iterated outputs
for two parse_expr examples

bertiewooster committed:
Remove outdated note to self

bertiewooster committed:
Improved formatting

bertiewooster commented:
I get a Sphinx build error `make: *** [Makefile:62: html] Error 1` when I add the "Solve an equation algebraically" sub-page to the solving folder. This occurs even if the sub-page's content is trivial, for example "Hi". So I'm guessing Sphinx is complaining that it's not been told where to put that sub-page. I tried a toctree directive for Markdown

````{toctree}`

(that's supposed to the three backticks followed by `{toctree}`) and still get a Sphinx error. Any suggestions?

bertiewooster committed:
Trying to include sub-page using toctree directive
on main page; Sphinx build fails

oscarbenjamin commented:
I can't immediately see the build logs because sphinx hasn't run due to doctest failures.

bertiewooster committed:
Fixing doctests

bertiewooster commented:
Got it. I pushed a commit that will hopefully fix the doctests, but I couldn't remember how to run doctests on my computer, so there may still be errors.

Could someone please remind me how to run doctests on my computer? I tried using this format

```./bin/doctest sympy/core/relational.py```

but got a `ModuleNotFoundError`, and I seem to recall that this isn't the way to run doctests.

bertiewooster committed:
Link from index to sub-page
Reformat solve() approaches to provide better table of contents
in Sphinx (Furo) right navbar

bertiewooster committed:
Minor edits

bertiewooster commented:
@asmeurer I have a note from our meeting, about solveset: the solution set can be more difficult to parse programmatically *trig function as example*. Could you please remind me what you meant by that?

bertiewooster committed:
Improve heading titles

bertiewooster commented:
Good news:

- doctests all pass
- [sub-page builds](https://output.circle-artifacts.com/output/job/e8dd1450-db86-4aa2-aa94-9350d0eaa4b6/artifacts/0/doc/_build/html/guides/solving/solve-equation-algebraically.html) and can be reached by clicking Solve an equation algebraically on the [solving main page
](https://output.circle-artifacts.com/output/job/e8dd1450-db86-4aa2-aa94-9350d0eaa4b6/artifacts/0/doc/_build/html/guides/solving/index.html)

Bad news:
- [Sphinx build still fails, reporting](https://github.com/sympy/sympy/runs/7058455628?check_suite_focus=true):

    ```
    build finished with problems, 1 warning.
    make: *** [Makefile:62: html] Error 1
    Error: Process completed with exit code 2.
    ```  
  Again, this makes me think that it's the inclusion of the sub-page in the folder, or under the main solving page, that's the problem.

oscarbenjamin commented:
The warning is visible further up in the log:
```
/home/runner/work/sympy/sympy/doc/src/guides/solving/solve-equation-algebraically.md:
WARNING: document isn't included in any toctree
```

bertiewooster committed:
Flesh out toctree:
hide
list sub-page

bertiewooster commented:
Great, thanks! That's what I suspected, and I didn't know how to build a toctree in Markdown. Turns out it's very similar to rst:

```
```{toctree}
:hidden: true

solve-equation-algebraically.md
(three backticks)
```

Sphix build succeeds locally now, so hopefully it will on the server too.

asmeurer commented:
> I don't immediately have one off hand but it should really be considered a bug. If there isn't an algorithm to find the solution then NotImplementedError should ideally be raised:

Or an example where it gives some solutions but not all. 

Actually, I find the NotImplementedError to be problematic for that reason:

```
>>> solve((cos(x) - x)*(x - 1))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/./sympy/solvers/solvers.py", line 1117, in solve
    solution = _solve(f[0], *symbols, **flags)
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/./sympy/solvers/solvers.py", line 1423, in _solve
    soln = _solve(m, symbol, **flags)
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/./sympy/solvers/solvers.py", line 1749, in _solve
    raise NotImplementedError('\n'.join([msg, not_impl_msg % f]))
NotImplementedError: multiple generators [x, cos(x)]
No algorithms are implemented to solve equation -x + cos(x)
```

Here it could have given the solution `[1]` even though it can't find the other solutions.

bertiewooster commented:
Understood. So it seems like that should that be reported as a bug, and not be in the content of this page?

asmeurer commented:
Yes. I think it is still possible for it to give some but not all of the solutions, but I'm not sure right now how to come up with an example.

oscarbenjamin commented:
> Here it could have given the solution `[1]` even though it can't find the other solutions.

There should be a different function for finding some solutions e.g. Mathematica has `FindInstance`:
https://reference.wolfram.com/language/ref/FindInstance.html

> I think it is still possible for it to give some but not all of the solutions,

A common case is when there are infinitely many solutions:
```python
In [14]: solve(sin(x), x)
Out[14]: [0, π]

In [15]: solve(x*exp(x)-1, x) # (infinitely many complex solutions)
Out[15]: [W(1)]
```

smichr commented:
> an example where a mathematical solution exists to an equation, but SymPy returns an empty list

see [here](https://github.com/sympy/sympy/issues/23013)

bertiewooster committed:
Add comment and link about transformations for parse_expr
Add example of solveset returning infinitely many solutions when solve does not
Updated equation on index.md to same as on sub-page

bertiewooster committed:
Fix indenting of a code block so it should render correctly

bertiewooster committed:
Fix doctest using import pprint

bertiewooster committed:
Changed prettyprint result to fix doctest

bertiewooster committed:
Add three more analogous examples for solveset
similar to solve examples

bertiewooster committed:
Cleanup

bertiewooster commented:
I believe this is all the planned content, so assuming the remaining tests pass, this is ready for review.

This page will serve as a model for the other Solving sub-pages.

oscarbenjamin commented on doc/src/guides/solving/solve-equation-algebraically.md:
```diff
@@ -0,0 +1,279 @@
+- {func}`~.solveset`
+    - You want a consistent input and output interface.
+    - You want to get all the solutions, including if there are infinitely many.
+    - You want a clear separation between equations in the complex domain and the real domain.
```
Note that solve can also be used to find real solutions:
```python
In [1]: x = symbols('x')

In [2]: solve(x**4 - 1, x)
Out[2]: [-1, 1, -ⅈ, ⅈ]

In [3]: x = symbols('x', real=True)

In [4]: solve(x**4 - 1, x)
Out[4]: [-1, 1]
```

oscarbenjamin replied:
Oh, I see you have that example below. What exactly does this statement about a "clear separation" mean?

bertiewooster replied:
Clarified how solveset limits the domain of results.

oscarbenjamin commented on doc/src/guides/solving/solve-equation-algebraically.md:
```diff
@@ -0,0 +1,279 @@
+    - You want to get all the solutions, including if there are infinitely many.
+    - You want a clear separation between equations in the complex domain and the real domain.
+
+We recommend you include the variable to be solved for as the second argument for either function. While this is optional for equations with a single symbol, it is a good practice because it ensures SymPy will solve for the desired symbol. For example, you may expect the following to solve for $x$, and SymPy will solve for $y$:
```
Long lines like this should wrapped in your editor.

oscarbenjamin commented on doc/src/guides/solving/solve-equation-algebraically.md:
```diff
@@ -0,0 +1,279 @@
+# Solve an equation algebraically
+
+Use SymPy to solve an equation algebraically (symbolically). For example, solving $x^2 = y$ yields $x \in \{-\sqrt{y},\sqrt{y}\}$.
```
This should immediately have links to other pages e.g. for when you have more than one equation or want to find integer solutions etc.

bertiewooster replied:
Linked to solving main page.

oscarbenjamin commented on doc/src/guides/solving/solve-equation-algebraically.md:
```diff
@@ -0,0 +1,279 @@
+
+### Make your equation into an expression that equals zero
+
+Use the fact that any expression not in an `Eq` (equation) is automatically assumed to equal zero (0) by the solving functions. You can rearrange the equation $x^2 = y$ to $x^2 - y = 0$, and {func}`~.solve` that expression. This approach is convenient if you are interactively solving an equation which can easily be rearranged to $expression = 0$.
```
It's pretty trivial to turn an equation into an expression:
```python
In [9]: eq = Eq(x, y)

In [10]: eq.lhs - eq.rhs
Out[10]: x - y

In [11]: eq.rewrite(Add)
Out[11]: x - y
```
The reason for not doing this is if it would cancel a singularity:
```python
In [12]: eq = Eq(x + 1/x, 1/x)

In [13]: eq
Out[13]: 
    1   1
x + ─ = ─
    x   x

In [14]: eq.rewrite(Add)
Out[14]: x

In [15]: eq.rewrite(Add, evaluate=False)
Out[15]: 
    1   1
x - ─ + ─
    x   x
```
I think many users expect that the input to `solve` should be equations rather than just expressions. This point about `Eq` should be made right at the beginning perhaps the first example should be with an equation.

Another terminology for understanding what it means to set an expression equal to zero is that we want to "find roots" of functions.

asmeurer replied:
The singularity thing seems like a very technical difference that probably isn't worth mentioning here.

asmeurer replied:
For most users, it comes down to whether they more naturally think of their problem as an equality or as an expression that equals 0. Also, from personal experience, I often use Eq just because it's easier to type than negating the entire right-hand side of an expression, and also less error prone (I might mess up minus sign with `-rhs`).

oscarbenjamin replied:
It does but from experience I can say that a lot of people find the idea of thinking of an expression as an equation implicitly set to zero unintuitive so this is something to be discussed early on.

Also the example should output what the Eq looks like so that it's clear(er) what a symbolic equality is.

oscarbenjamin commented on doc/src/guides/solving/solve-equation-algebraically.md:
```diff
@@ -0,0 +1,279 @@
+
+Parse a string representing the equation into a form that SymPy can understand (`Eq` form), then apply {func}`~.solve` to the parsed expression.  This approach is convenient if you are programmatically reading in a string. We [recommend against using parsing a string if you are creating the expression yourself](https://github.com/sympy/sympy/wiki/Idioms-and-Antipatterns#strings-as-input). Parsing ...
+
+You should always include the variable to solve for if you want to extract results programmatically, to ensure that SymPy solves for the desired variable. To ensure SymPy will produce results in a consistent format, use `dict=True`. To extract the solutions, you can iterate through the list of dictionaries:  
```
The other advantage of using `dict=True` is that it works nicely with `subs`. This is mentioned above but there are no examples. This is important because it is really the main purpose in having a `solve` function at all.

bertiewooster replied:
Added an example of `subs` using critical points.

oscarbenjamin commented on doc/src/guides/solving/solve-equation-algebraically.md:
````diff
@@ -0,0 +1,279 @@
+{2*n*pi | n in Integers} U {2*n*pi + pi | n in Integers}
+```
+
+However, {func}`~.solve` will return only a finite number of solutions (here, two):
+
+```py
+>>> from sympy import sin, solve
+>>> from sympy.abc import x
+>>> solution = solve(sin(x), x)
+>>> print(solution)
+[0, pi]
+```
````
This should clarify why it is only 2 solutions and how that choice is made. Essentially the idea is that all solutions are generated from these two by adding multiples of the periodicity of the equation (here `2*pi`). The periodicity of other examples can be different:
```
In [18]: print(solve([sin(x)+sin(x/2)], [x], dict=True))
[{x: 0}, {x: -8*pi/3}, {x: -2*pi}, {x: -4*pi/3}, {x: 4*pi/3}, {x: 2*pi}, {x: 8*pi/3}, {x: 4*pi}]
```
(Not sure if that's really correct output)

oscarbenjamin replied:
> Is this actually true in general?

Probably not but the I think the intention is for it to be true (CC @smichr).

bertiewooster replied:
Stated why two, and that this is the goal.

oscarbenjamin commented on doc/src/guides/solving/solve-equation-algebraically.md:
````diff
@@ -0,0 +1,279 @@
+No algorithms are implemented to solve equation -x + cos(x)
+```
+
+so you may have to {func}`solve your equation numerically <sympy.solvers.solvers.nsolve>` instead.
````
I think that this alternative suggestion should be presented immediately at the top of this page. For many users this point is the most important point. Right at the beginning we should clarify the distinction between analytic and numeric solutions.

asmeurer commented on doc/src/guides/solving/solve-equation-algebraically.md:
```diff
@@ -0,0 +1,279 @@
+## Using {func}`~.solve`
+
+- produces various output formats depending on the answer
+    - unless you use `dict=True` to ensure the result will be formatted as a dictionary, which we recommend if you want to extract information from the result programmatically
```
This is related to the discussion that is happening on the mailing list. To me, the second solution is not right. If I tell solve to solve for `[a, b]`, then I don't want `a` or `b` to be included in the solutions. Is this the only difference that occurs when passing a list for a single equation? This is the first time I've seen this advice.

oscarbenjamin replied:
> Is this the only difference that occurs when passing a list for a single equation? This is the first time I've seen this advice.

There could be a lot more cases. They go through completely different code paths and `_solve_system` does a lot before it falls back on `_solve`.

Library code usually doesn't use solve with a single expression so the list case is more well tested and fixed.

> To me, the second solution is not right. If I tell solve to solve for `[a, b]`, then I don't want `a` or `b` to be included in the solutions.

This is just how solve represents the solution of an underdetermined system:
```python
In [1]: solve([x + y + z, x - 1], [x, y, z])
Out[1]: {x: 1, y: -z - 1}

In [2]: linsolve([x + y + z, x - 1], [x, y, z])
Out[2]: {(1, -z - 1, z)}

In [3]: nonlinsolve([x + y + z, x - 1], [x, y, z])
Out[3]: {(1, -z - 1, z)}

In [4]: groebner([x + y + z, x - 1], [x, y, z])
Out[4]: GroebnerBasis([x - 1, y + z + 1], x, y, z, domain=ℤ, order=lex)

In [5]: linsolve(_, [x, y])
Out[5]: {(1, -z - 1)}
```
There are other ways but there needs to be a free parameter in the solution so it either needs to be one of the variables or it needs to be a new symbol like the `Ci` that `dsolve` uses to represent families of solutions.

The idea behind `solve` us fundamentally that you want an explicit representation of the solution set so that you can substitute it into to something else to eliminate the symbols given as unknowns. With an underdetermined system it isn't possible to get a substitution rule that will eliminate all unknowns but we can make a substitution rule that will eliminate a subset of the unknowns. The result after substituting is then given in terms of the free parameters which can be a subset of the unknowns or otherwise need to be new symbols. The choice made by `solve` is not to introduce new symbols.

bertiewooster replied:
Use dict=True in all solve() examples.

oscarbenjamin replied:
> Is this the only difference that occurs when passing a list for a single equation? This is the first time I've seen this advice.

See also here where it looks like a bug in heurisch can be fixed by passing a list to solve instead of a single equation:
https://github.com/sympy/sympy/issues/23718#issuecomment-1173733427

asmeurer commented:
> There should be a different function for finding some solutions e.g. Mathematica has FindInstance:
[reference.wolfram.com/language/ref/FindInstance.html](https://reference.wolfram.com/language/ref/FindInstance.html)

Well solve already returns only some of the solutions for the infinite solutions case. Equations like the above where part can be solved are not uncommon. For example, to finding the critical points of `f(g(x))` involves solving `f'(g(x))*g'(x)`. It can easily happen that `f'(g(x))` isn't solvable (in closed-form) but `g'(x)` is.

bertiewooster committed:
Text changes per reviews

bertiewooster committed:
Use dict=True in all solve() examples

bertiewooster committed:
Add simple example of solve() and solveset()
before comparing the functions

bertiewooster committed:
Specify to solve for x three times

bertiewooster committed:
Add subs() example of critical values

bertiewooster committed:
{meth}`~.subs` instead of {func}`~.subs`

bertiewooster committed:
Print critical point x and y values together

bertiewooster committed:
{meth}`~sympy.core.basic.Basic.subs`
using path to method because more than one subs() in SymPy

bertiewooster committed:
Change guides/index toc maxdepth to 1

bertiewooster commented:
I believe I addressed all the review comments.

I checked the table of contents (toc) listing on guides/index and for some reason the solving sub-page isn't included, just the solving main page. I haven't been able to figure out why. However, I think we should limit the guides toc to one level because there are 32 items listed already, and will be about a dozen more after this Season of Docs project, so it gets to be too much for visitors to skim through--more than a screen height on a laptop.

So I changed the guides toc `maxdepth` to 1, so only the top-level guides are listed. Let me know what you think.

bertiewooster committed:
Specify to solve for x
in equation with no solution

oscarbenjamin commented on doc/src/guides/solving/solve-equation-algebraically.md:
```diff
@@ -0,0 +1,394 @@
+Here are recommendations on when to use:
+
+- {func}`~.solve`
+    - You need to programmatically extract components (expressions, or individual symbols or constants) 
+    from the output by traversing it.
+    - You want an explicit solution expression that you can use to substitute into something else.
+    - You want to apply assumptions to symbols (for example, $x$ is real) to determine the 
+    domain of the solutions.
```
I don't think this last point really makes sense as a reason to use `solve`. I would probably just remove it. In fact I'm not sure about the first point either as written.

How about:
```
- You want to get explicit symbolic representations of the different values
  a variable could take that would satisfy the equation.
- You want to substitute those explicit solution values into other equations
  or expressions involving the same variable.
```

oscarbenjamin commented on doc/src/guides/solving/solve-equation-algebraically.md:
```diff
@@ -0,0 +1,394 @@
+- {func}`~.solveset`
+    - You want a consistent input and output interface.
+    - You want to get all the solutions, including if there are infinitely many.
+    - You want to limit the domain of the solutions (for example, to real values) directly.
```
I think many users would say that the output of `solveset` is not consistent. Really they want a FiniteSet but instead they have a Complement of an Intersection of a Union etc: how do I get the values out of this?

Also the part about limiting the domain of the solutions to the reals doesn't make much sense since solve also allows you to do that. What solveset allows you to do is to limit the domain of the solutions to any arbitrary set e.g.:
```python
In [1]: solveset(sin(x), x, Interval(-pi, pi))
Out[1]: {0, -π, π}
```

oscarbenjamin commented on doc/src/guides/solving/solve-equation-algebraically.md:
```diff
@@ -0,0 +1,394 @@
+>>> from sympy.abc import x
+>>> from sympy.parsing.sympy_parser import convert_equals_signs, standard_transformations
+>>> expr = "x ** 2 = y"
+>>> parsed = parse_expr(expr, transformations=standard_transformations + (convert_equals_signs,))
```
Probably simpler to use `transformations="all"`:
```python
In [2]: parse_expr("x**2 = y", transformations="all")
Out[2]: 
 2    
x  = y
```

bertiewooster replied:
Good tip. Thanks!

oscarbenjamin commented on doc/src/guides/solving/solve-equation-algebraically.md:
```diff
@@ -0,0 +1,394 @@
+>>> from sympy import solve, diff
+>>> f = x**3 + x**2 - x
+>>> derivative = diff(f, x)
+>>> critical_points = solve(derivative, x, dict=True)
+>>> print(critical_points)
+[{x: -1}, {x: 1/3}]
+>>> for critical_point in critical_points:
+...     # Extract the value from the dictionary where key is x
+...     critical_point_value = critical_point[x]
+...     f_at_critical_point = f.subs(x, critical_point_value)
+...     print(f"f({critical_point_value}) = {f_at_critical_point}")
+f(-1) = 1
+f(1/3) = -5/27
```
This is more complicated than it needs to be. I would just show this as:
```python
In [5]: critical_points = solve(derivative, x, dict=True)

In [6]: critical_points
Out[6]: [{x: -1}, {x: 1/3}]

In [7]: point1, point2 = critical_points

In [8]: f.subs(point1)
Out[8]: 1

In [9]: f.subs(point2)
Out[9]: -5/27
```
The point is that a dict can be used directly with `subs`. That's why a list of dicts is the best output because the main purpose of `solve` is that you will substitute the solution into something.

oscarbenjamin commented on doc/src/guides/solving/solve-equation-algebraically.md:
```diff
@@ -0,0 +1,394 @@
+- produces outputs in the format of 
+[SymPy mathematical Sets](https://docs.sympy.org/dev/modules/sets.html?highlight=sets#module-sympy.sets.sets) rather than 
+[Python sets](https://docs.python.org/3/library/stdtypes.html#set)
+- can return infinitely many solutions
```
I would say something more like "can represent infinite sets of possible solutions"

oscarbenjamin commented on doc/src/guides/solving/solve-equation-algebraically.md:
```diff
@@ -0,0 +1,394 @@
+### Parse a string representing the equation
+Parse a string representing the equation into a form that SymPy can understand (`Eq` form), then apply 
+{func}`~.solveset` to the parsed expression.  This approach is convenient if you are programmatically 
+reading in a string. We [recommend against using parsing a string if you are creating the expression 
```
I don't think this needs to be repeated.

oscarbenjamin replied:
Same for "eq form" above.

bertiewooster replied:
By "this" do you mean just the narrative preambles, or also the code examples?

oscarbenjamin replied:
The whole section. There is a section above about parsing expressions and using that with `solve`. This section is basically a repetition of that. Same goes for "put your equation in Eq form". We don't need to duplicate these but rather can note once that an equation parsed from a string can be passed to either solve or solveset and both accept Eq or Expr.

bertiewooster replied:
I had considered reorganizing the guide to be by solving approach (expression, `Eq`, or parsing) first and then by type (`solve` or `solveset`). I did that to address your feedback, make the guide shorter, and let visitors easily compare the output of `solve` or `solveset`.

oscarbenjamin commented on doc/src/guides/solving/solve-equation-algebraically.md:
```diff
@@ -0,0 +1,394 @@
+>>> from sympy.abc import x, y
+>>> solution = solveset(x**2 - y, x)
+>>> print(solution)
+{-sqrt(y), sqrt(y)}
```
Question number 1 to be answered here is how do I actually get the expressions out of this object?
```
In [11]: solution[0]
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-11-ed0f89115762> in <module>
----> 1 solution[0]

TypeError: 'FiniteSet' object is not subscriptable
```

bertiewooster replied:
Good question. I learned that iterating through a finite set returns its expressions.

oscarbenjamin replied:
But you don't always get a FiniteSet:
```python
In [1]: x, y = symbols('x, y')

In [2]: solveset(x**2 - y, x, Reals)
Out[2]: ℝ ∩ {-√y, √y}

In [3]: list(_)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-1a5b60e28d67> in <module>
----> 1 list(_)

~/current/sympy/sympy.git/sympy/sets/sets.py in __iter__(self)
   1554             if not candidates:
   1555                 raise TypeError("None of the constituent sets are iterable")
-> 1556             raise TypeError(
   1557                 "The computation had not completed because of the "
   1558                 "undecidable set membership is found in every candidates.")

TypeError: The computation had not completed because of the undecidable set membership is found in every candidates.
```
The question is what to do? Here if you declare `y` to be positive you'll get the FiniteSet:
```python
In [4]: y = symbols('y', positive=True)

In [5]: solveset(x**2 - y, x, Reals)
Out[5]: {-√y, √y}
```
It's not always possible in general to set appropriate assumptions though so the question is how do you detect if solveset returned a FiniteSet and what do you do about it when it doesn't? This is actually one of the main issues that users have with using `solveset` and also the main reason that many users prefer `solve` so I think this guide needs to address it.

oscarbenjamin commented on doc/src/guides/solving/solve-equation-algebraically.md:
````diff
@@ -0,0 +1,394 @@
+>>> from sympy import solve, Eq
+>>> from sympy.abc import x
+>>> eqn = Eq(x - 7, x + 2)
+>>> solve(eqn, x)
+[]
+```
+
+So if SymPy returns an empty set, you may want to check whether there is a mistake in the equation.
````
This is an empty list rather than an empty set. It's important to be clear here because `solveset` returns the mathematical empty set.

oscarbenjamin commented on doc/src/guides/solving/solve-equation-algebraically.md:
````diff
@@ -0,0 +1,394 @@
+{-4, 4, -4*I, 4*I}
+```
+
+If you want to restrict returned solutions to real numbers, you can specify the 
+[domain to solve in](https://docs.sympy.org/dev/modules/solvers/solveset.html?highlight=solveset#what-is-this-domain-argument-about) 
+as `S.Reals`:
````
Showing how to specify something other than Reals might also be useful here.

oscarbenjamin commented:
In the case of `solve` something that comes up on SO a lot is people using either `check=False` or `simplify=False`. We should clarify what those do here and why someone might want to use them. Setting either to False certainly speeds things up but setting `check` to False gives potentially incorrect results.

bertiewooster committed:
Updates per 2022-07-02 review

bertiewooster committed:
Futher improvements

bertiewooster committed:
Merge branch 'sympy:master' into equation-algebraically

bertiewooster committed:
Added references to solveset.rst, linked to

bertiewooster committed:
Merge branch 'equation-algebraically' of github.com:bertiewooster/sympy into equation-algebraically

bertiewooster committed:
Reorganize by approach, solve and solveset
within each approach

bertiewooster committed:
Tweak organization
Fix three failed doctests

bertiewooster committed:
Fix doctest designed to throw an error

bertiewooster committed:
Fix take 2: doctest designed to throw an error

bertiewooster committed:
Fix take 3: doctest designed to throw an error

smichr commented on doc/src/guides/solving/solve-equation-algebraically.md:
````diff
@@ -0,0 +1,379 @@
+solution = solve(x ** 4 - 256, x, dict=True)
+print(solution)
+[{x: -4}, {x: 4}]
+```
````
You can also include inequalities to limit variables to real values:
```python
>>> solve(((x**2+2)*(x-3)*(x-2)*(x-1),x>0,x<3))
Or(Eq(x, 1), Eq(x, 2))
```
Including only `x>0` will then include `Eq(x, 3)`. But you can't currently inlcude discontinuous ranges since something like `Or(x<0,x>2)` will raise an error.

smichr commented:
You might want to check my summary of current input/output types [here](https://github.com/sympy/sympy/pull/23699#discussion_r913355272).

smichr commented on doc/src/guides/solving/solve-equation-algebraically.md:
````diff
@@ -0,0 +1,379 @@
+```
+
+{func}`~.solve` tries to return just enough solutions so that all (infinitely many) solutions can generated 
+from the returned solutions by adding integer multiples of the periodicity of the equation, here $2\pi$.
````
Would it be worth showing the use of the `periodicity` function here?

smichr commented on doc/src/guides/solving/solve-equation-algebraically.md:
````diff
@@ -0,0 +1,379 @@
+1
+>>> print(f.subs(point2))
+-5/27
+```
````
Maybe reinforce multiple uses of info?
```
You can tell if the critical point is a maxima or minima by substituting the values into the expression for the second derivative:

>>> curvature = diff(f, x, 2)
>>> curvature.subs(point1)
```
 etc...

bertiewooster committed:
Add another substitution example:
second derivative negative for maximum, positive for minimum

bertiewooster committed:
Demonstrate periodicity function

bertiewooster committed:
Add other ways to restrict solve() domain
Fix Derivative reference: class, not function

bertiewooster commented:
> You might want to check my summary of current input/output types [here](https://github.com/sympy/sympy/pull/23699#discussion_r913355272).

Thanks @smichr. One thing I noticed is that `solve` ignores `dict=True` if passed a boolean expression:
```
>>> from sympy import Symbol, solve
>>> x = Symbol('x', real=True)
>>> expr = (x-4)*(x-3)*(x-2)*(x-1)
>>> solution = solve(expr, x, dict=True)
>>> print(solution)
[{x: 1}, {x: 2}, {x: 3}, {x: 4}]
>>> solution_2_3 = solve((expr,x>=2,x<=3), x, dict=True)
>>> print(solution_2_3)
Eq(x, 2) | Eq(x, 3)
```

bertiewooster committed:
Fix doctests by removing indentation

oscarbenjamin commented:
> `solve` ignores `dict=True` if passed a boolean expression:

The behaviour of solve with inequalities is completely different from how it behaves normally. The guide should clarify that including inequalities to filter solutions does not really do what many people would expect e.g.:
```python
In [3]: solve([x*y - 1, y - 1], [x, y])
Out[3]: [(1, 1)]

In [4]: solve([x*y - 1, y - 1, x > 1], [x, y])
---------------------------------------------------------------------------
NotImplementedError
```
There's just a completely different codepath in solve that does this and it can only handle univariate inequalities so even trivial cases like the one above don't work as someone might expect.

oscarbenjamin commented on doc/src/guides/solving/solve-equation-algebraically.md:
```diff
@@ -0,0 +1,408 @@
+
+So if SymPy returns an empty list, you may want to check whether there is a mistake in the equation.
+
+### Equations with no analytical solution
+
+The vast majority of arbitrary nonlinear equations are not analytically solvable. 
+The classes of equations that are solvable are basically:
+1. Linear equations
+2. Polynomials, except where limited by the Abel-Ruffini theorem
+3. Equations that can be solved by inverting some transcendental functions
+4. Problems that can be transformed into the cases above 
+(e.g., by turning trigonometric functions into polynomials)
+5. A few other special cases that can be solved with something like the Lambert W function
```
This topic probably deserves its own dedicated guide.

bertiewooster replied:
That's fair. I'm thinking let's leave this in this page for now until this pull request is merged, then I'll split it out into its own page "Which kinds of equations can be solved".

bertiewooster committed:
Fix doctests

bertiewooster commented:
> > `solve` ignores `dict=True` if passed a boolean expression:
> 
> The behaviour of solve with inequalities is completely different from how it behaves normally. The guide should clarify that including inequalities to filter solutions does not really do what many people would expect e.g.:
> 
> ```python
> In [3]: solve([x*y - 1, y - 1], [x, y])
> Out[3]: [(1, 1)]
> 
> In [4]: solve([x*y - 1, y - 1, x > 1], [x, y])
> ---------------------------------------------------------------------------
> NotImplementedError
> ```
> 
> There's just a completely different codepath in solve that does this and it can only handle univariate inequalities so even trivial cases like the one above don't work as someone might expect.

Good to know. Let's save this quirk for the guide on solving sets of equations--I don't want to introduce here the complication of passing `solve` an argument that's a _list_ of expressions.

bertiewooster committed:
Explain check=False and simplify=False

bertiewooster committed:
Clarify headings

bertiewooster committed:
Minor edits

bertiewooster commented:
> speed

Added a section **Options that can speed up solve**

bertiewooster committed:
Minor edits

bertiewooster commented:
I believe I have addressed all the comments. I'm pretty sure the two optional-dependencies failures of test_aesaracode is unrelated to this pull request, so request final review and merging.

oscarbenjamin commented on doc/src/guides/solving/solve-equation-algebraically.md:
```diff
@@ -0,0 +1,449 @@
+At the critical points, the {class}`~.Derivative` equals zero (or is undefined). 
+You can then obtain the function values at those critical points
+by substituting the critical points back into the function using {meth}`~sympy.core.basic.Basic.subs`. You can also tell if the critical point is a maxima or minima by substituting the values
+into the expression for the second derivative: a negative value indicates a maximum, and a positive value indicates a minimum.
```
Long lines should be wrapped. This is something that I do with my editor. I don't know what editor you use but I use vim and I set the textwidth to 78. Then I can highlight a paragraph like this use `:gq` to wrap the lines to that width. Other editors will have ways of doing the same thing.

bertiewooster replied:
I'm using Visual Studio Code and just installed the ReWrap extension to do that by pressing Option-Q on my Mac (Alt-Q on Windows).

oscarbenjamin commented:
I think it would be good to mention evalf and how to use that for purely numeric solutions that are complicated or not obvious e.g.:
```python
In [17]: solutions = solve(x**5 - x - 1, x, dict=True)

In [18]: solutions
Out[18]: 
⎡⎧          ⎛ 5           ⎞⎫  ⎧          ⎛ 5           ⎞⎫  ⎧          ⎛ 5           ⎞⎫  ⎧          ⎛ 5 
⎢⎨x: CRootOf⎝x  - x - 1, 0⎠⎬, ⎨x: CRootOf⎝x  - x - 1, 1⎠⎬, ⎨x: CRootOf⎝x  - x - 1, 2⎠⎬, ⎨x: CRootOf⎝x  
⎣⎩                         ⎭  ⎩                         ⎭  ⎩                         ⎭  ⎩              

          ⎞⎫  ⎧          ⎛ 5           ⎞⎫⎤
- x - 1, 3⎠⎬, ⎨x: CRootOf⎝x  - x - 1, 4⎠⎬⎥
           ⎭  ⎩                         ⎭⎦

In [19]: [s[x].n(3) for s in solutions]
Out[19]: [1.17, -0.765 - 0.352⋅ⅈ, -0.765 + 0.352⋅ⅈ, 0.181 - 1.08⋅ⅈ, 0.181 + 1.08⋅ⅈ]

```
As an aside I just saw another weirdness for list vs non-list:
```python
In [20]: solve([x**5 - x - 1], x)
Out[20]: []

In [21]: solve(x**5 - x - 1, x)
Out[21]: 
⎡       ⎛ 5           ⎞         ⎛ 5           ⎞         ⎛ 5           ⎞         ⎛ 5           ⎞        
⎣CRootOf⎝x  - x - 1, 0⎠, CRootOf⎝x  - x - 1, 1⎠, CRootOf⎝x  - x - 1, 2⎠, CRootOf⎝x  - x - 1, 3⎠, CRootO

 ⎛ 5           ⎞⎤
f⎝x  - x - 1, 4⎠⎦
```
The non-list result is the preferred result here but in any case they should return the same solutions.

oscarbenjamin commented on doc/src/guides/solving/solve-equation-algebraically.md:
````diff
@@ -0,0 +1,449 @@
+[sqrt(y), -sqrt(y)]
+```
+
+However, for more complex results, it may not be possible to list the solutions:
````
This should explain both why `solveset` returned an unevaluated `Intersection` and also how to avoid this:

- Declare `y` to be real and positive
- Extract the FiniteSet from the Intersection.

bertiewooster replied:
>why solveset returned an unevaluated Intersection

>Extract the FiniteSet from the Intersection

These would be good to include, and are beyond my SymPy knowledge, nor could I find them in the documentation. Could you point me to, or give, an explanation/example as a basis for me to document?

bertiewooster replied:
I worked up an example that I believe addresses this suggestion.

oscarbenjamin replied:
To get the FiniteSet from an intersection you can use `.args`.

bertiewooster committed:
Add evalf.n to get numerical solutions

bertiewooster commented:
> I think it would be good to mention evalf and how to use that for purely numeric solutions that are complicated or not obvious e.g.:

That would be useful for non-closed form solutions. I'm adding.

bertiewooster commented on doc/src/guides/solving/solve-equation-algebraically.md:
````diff
@@ -0,0 +1,449 @@
+```py
+>>> from sympy import parse_expr, solve, solveset
+>>> from sympy.abc import x
+>>> expr = "x ** 2 = y"
````
Changed exponential operator from `**` to `^` and touted the latter as a more mathematical notation.

bertiewooster committed:
Demonstrate can parse ^ as exponent operator
Demonstrate eqn object created

bertiewooster committed:
Wrap long lines

bertiewooster committed:
Remove many print statements from code blocks

bertiewooster commented on doc/src/guides/solving/solve-equation-algebraically.md:
```diff
@@ -0,0 +1,449 @@
+>>> from sympy import solveset
+>>> from sympy.abc import x, y
+>>> solution = solveset(x**2 - y, x)
+>>> print(solution)
```
I thought each code block was tested in a separate context, so I thought each had to import all packages it used. But now, as I understand it, the whole page is tested in a single context, so that's not necessary. As a new user of a package, though, I appreciate being able to simply click the copy button on a code block, paste it into my editor, and know that it will run--it can be difficult to figure out which sub-module to import from. So I think it's worth repeating the couple of lines to make it easier on new users. But if the community consensus is to remove repetitive imports, I can do that.

Similarly, I thought I had to use `print` statements because outputs aren't produced in my development environment from e.g. 

```py
>>> solveset(x**2 - y, x)
```

But I understand that doctest will print such results, so I've removed many `print` statements.

oscarbenjamin replied:
> outputs aren't produced in my development environment from e.g.
> 
> ```python
> >>> solveset(x**2 - y, x)
> ```

What "development environment" are you using? If you run that in any ordinary Python REPL you will see some output.

bertiewooster replied:
I'm using Visual Studio Code. If I run a script there in Run and Debug mode, or execute the script from the terminal, it does not produce output on such steps, which I'll term implicit output. If I run in the IDLE Shell, it does produce implicit output. So I appreciate your point that doctests run more like a REPL than a script.

I find it easier to put the code in a file and then run it to ensure that the code in the script has everything it needs (imports) to run. Contrarily, if I develop and run code interactively in a REPL, I might not copy into the docstring some line that I forgot was needed to make the code work.

I discovered that VS Code has interactive commands where you can right-click code in a file and select e.g. Run Current File in Interactive Window. That creates an interactive window which behaves like to a Jupyter Notebook. It works pretty well for this purpose; a drawback is that it only produces implicit output from the last line in the script. It is possible to produce implicit output from a previous line by right-clicking on it and selecting Run Selection/Line in Interactive Window. So I'll plan to use those features from now on.

asmeurer replied:
Repeating imports is fine when it's not clear that the person reading the example will have read the previous examples where the original imports are. They're not needed but they can help. I do agree we should just use `solveset(x**2 - y, x)` and let it show the output instead of `print(solution)`.

bertiewooster replied:
Thanks. I did remove the unnecessary `print` statements, now that I understand doctests are run in a REPL.

bertiewooster committed:
Removed two redundant sentences

bertiewooster commented on doc/src/guides/solving/solve-equation-algebraically.md:
```diff
@@ -0,0 +1,449 @@
+You should always include the variable to solve for if you want to extract results programmatically, 
+to ensure that SymPy solves for the desired variable. 
+To ensure SymPy will produce results in a consistent format, use `dict=True`. 
+To extract the solutions, you can iterate through the list of dictionaries:  
```
Removed two redundant sentences.

smichr commented:
> As an aside I just saw another weirdness for list vs non-list:

I'm guessing this is because `solve_poly_system` does not allow for non-radical solutions of equations.

smichr commented:
A troublesome behavior of `solve` that will be phased out is the automatic detection of undetermined coefficients. I added an update to the docstring in #23699. Is it worth mentioning here: mostly, just how to make that request to `solve_undetermined_coeffs` directly rather than using `solve` to do it? `solve` will solve as undetermined coeff system when a single, multivariate equation and all but one of the variables is specified, e.g. `solve(a*x**2 + b - 4, a, b)` gives `{a: 0, b: 4}` and changing `b` to `b**2` gives [(0, -2), (0, 2)]`.

Is this going to go in for 1.11? If so, it would be nice to get #23699 in. And rather than typing `dict=True` one could use
```python
    from sympy import solve as __solve
    def solve(a, *b, **c):
        from sympy.utilities.iterables import iterable
        if b:
            if len(b) == 1 and iterable(b[0]):
                return __solve(a, b[0], **c, dict=True)
            return __solve(a, *b, **c, dict=True)
        return __solve(a, **c, dict=True)
    del __solve
```
of define once, `D=dict(dict=True)` then use `solve(x - 1, **D)`.

bertiewooster committed:
Demonstrate how to resolve intersection
of solution set and domain

bertiewooster committed:
Add example of extracting sets
from Intersection(set1, set2) using .args

bertiewooster committed:
Trying role=any for args

asmeurer commented:
Undetermined coefficients could be its own guide but I don't think it needs to be mentioned in this one.

bertiewooster committed:
Fix typo

bertiewooster committed:
Add example of using nsolve

bertiewooster committed:
Remove "Parse a string representing the equation"
Will move into solving guidance page

bertiewooster committed:
Remove"Equations with no analytical solution" section
Will move into solving guidance

asmeurer commented:
Looks like all the comments here have been addressed, so I am going to merge this. If there are any additional comments, they can be fixed in a follow up PR.

oscarbenjamin commented on doc/src/guides/solving/solve-equation-algebraically.md:
```diff
@@ -0,0 +1,413 @@
+[1, 4]
+>>> solution_2_3 = solve((expr,x>=2,x<=3), x)
+>>> print(solution_2_3)
+Eq(x, 2) | Eq(x, 3)
```
We really shouldn't recommend this as a way of using `solve`. Just look at the weird output that it generates. This is almost certainly not what someone would have expected:
```python
In [1]: eq = (x-4)*(x-3)*(x-2)*(x-1)

In [2]: solve(eq, x)
Out[2]: [1, 2, 3, 4]

In [3]: solve((eq, x>=2, x<=3), x)
Out[3]: x = 2 ∨ x = 3
```
If we're going to show things like this then we need to explain why this happens. Really I would list this under a "caveats" section rather than as a tip for how to use solve.

bertiewooster replied:
Removed this recommendation in #23805 due to confusing output formatting

oscarbenjamin commented on doc/src/guides/solving/solve-equation-algebraically.md:
```diff
@@ -0,0 +1,413 @@
+>>> solve(x ** 2 - y, x, dict=True)
+[{x: -sqrt(y)}, {x: sqrt(y)}]
+>>> solveset(x**2 - y, x)
+{-sqrt(y), sqrt(y)}
```
In many places code snippets are inconsistent in the spacing around `**`. I prefer no spaces like `x**2` but either way it should be consistent.

bertiewooster replied:
Removed spaces around `**` so is like `x**2` in #23805
